### DISCUSSION TITLE
Is there a way to save sessions automatically?

### DISCUSSION THREAD

**Discussion by rieje** [jarun/nnn#1656] [Q&A]
Is there a way to save sessions automatically? E.g. if the terminal closes with nnn running then relaunching the session will restore it to its latest state instead of the last saved state (i.e. when nnn was explcitly closed with a `Q`) as it does currently. Useful to avoid the need to go to each nnn instance and `Q` before shutting down the system to save the sessions, which is annoying when you're managing multiple nnn instances in a Tmux session.

I was thinking maybe the state of the session gets updated in tmpfs and then regardless of how nnn gets terminated it will be written to disk assuming you're using `-S` to enable persistent sessions. Then it should be possible to just shutdown the system with sessions saved automatically without remembering to go to each instance and `Q`.

**Comment by jarun** [comment]
`-S` already does that. Read more about sessions in wiki.

**Reply by rieje** [reply]
I'm using `-S` and `-s` already, it works as expected with `Q` to exit nnn but not when nnn is terminated by closing the terminal window while nnn is running. I.e. according to manual, "program quit" refers to quitting with `Q` but not when nnn quits from closing terminal instance running nnn.

I was also thinking nnn accept a signal to tell it to save sessions and close.

**Reply by N-R-K** [reply]
> I was also thinking nnn accept a signal to tell it to save sessions and close

Signal handler are [*very tricky*](https://man7.org/linux/man-pages/man7/signal-safety.7.html) to get right.

In fact, current `nnn`'s signal handlers aren't safe because it calls `exit` (which flushes `stdio` streams and calls `atexit` - which are not safe to do from a signal handler).

**Comment by jarun** [comment]
Sessions is a long-supported feature and has not been changed in recent past.

I tried the following option on master and don't see any issues:

    nnn -S -s hello

I was using `q` to quit.
